File: src/app/aiRuntimeStats.js
function clampNonNegative(value, fallback = 0) {
  const safe = Number(value);
  if (!Number.isFinite(safe)) return fallback;
  return Math.max(0, safe);
}

export function createDefaultAiRuntimeStats() {
  return {
    requestCount: 0,
    responseCount: 0,
    environmentRequests: 0,
    policyRequests: 0,
    environmentResponses: 0,
    policyResponses: 0,
    llmResponseCount: 0,
    fallbackResponseCount: 0,
    timeoutCount: 0,
    errorCount: 0,
    recoveryCount: 0,
    avgLatencyMs: 0,
    lastLatencyMs: 0,
    lastRequestSec: -999,
    lastResponseSec: -999,
    lastLiveSec: -999,
    lastFallbackSec: -999,
    maxUnrecoveredFallbackSec: 0,
    consecutiveFallbackResponses: 0,
    lastErrorKind: "none",
    lastErrorMessage: "",
    lastResultSource: "none",
    coverageTarget: "fallback",
    liveCoverageSatisfied: false,
  };
}

export function getAiCoverageTarget(state) {
  const raw = String(state?.ai?.coverageTarget ?? "").trim().toLowerCase();
  if (raw === "llm") return "llm";
  return "fallback";
}

export function ensureAiRuntimeStats(state) {
  state.metrics.aiRuntime ??= createDefaultAiRuntimeStats();
  const stats = state.metrics.aiRuntime;
  stats.requestCount = clampNonNegative(stats.requestCount);
  stats.responseCount = clampNonNegative(stats.responseCount);
  stats.environmentRequests = clampNonNegative(stats.environmentRequests);
  stats.policyRequests = clampNonNegative(stats.policyRequests);
  stats.environmentResponses = clampNonNegative(stats.environmentResponses);
  stats.policyResponses = clampNonNegative(stats.policyResponses);
  stats.llmResponseCount = clampNonNegative(stats.llmResponseCount);
  stats.fallbackResponseCount = clampNonNegative(stats.fallbackResponseCount);
  stats.timeoutCount = clampNonNegative(stats.timeoutCount);
  stats.errorCount = clampNonNegative(stats.errorCount);
  stats.recoveryCount = clampNonNegative(stats.recoveryCount);
  stats.avgLatencyMs = clampNonNegative(stats.avgLatencyMs);
  stats.lastLatencyMs = clampNonNegative(stats.lastLatencyMs);
  stats.lastRequestSec = Number.isFinite(Number(stats.lastRequestSec)) ? Number(stats.lastRequestSec) : -999;
  stats.lastResponseSec = Number.isFinite(Number(stats.lastResponseSec)) ? Number(stats.lastResponseSec) : -999;
  stats.lastLiveSec = Number.isFinite(Number(stats.lastLiveSec)) ? Number(stats.lastLiveSec) : -999;
  stats.lastFallbackSec = Number.isFinite(Number(stats.lastFallbackSec)) ? Number(stats.lastFallbackSec) : -999;
  stats.maxUnrecoveredFallbackSec = clampNonNegative(stats.maxUnrecoveredFallbackSec);
  stats.consecutiveFallbackResponses = clampNonNegative(stats.consecutiveFallbackResponses);
  stats.lastErrorKind = String(stats.lastErrorKind ?? "none");
  stats.lastErrorMessage = String(stats.lastErrorMessage ?? "");
  stats.lastResultSource = String(stats.lastResultSource ?? "none");
  stats.coverageTarget = getAiCoverageTarget(state);
  stats.liveCoverageSatisfied = Boolean(stats.liveCoverageSatisfied);
  return stats;
}

export function resetAiRuntimeStats(state) {
  state.metrics.aiRuntime = createDefaultAiRuntimeStats();
  state.metrics.aiRuntime.coverageTarget = getAiCoverageTarget(state);
  return state.metrics.aiRuntime;
}

export function classifyAiErrorMessage(errorText = "") {
  const raw = String(errorText ?? "").trim().toLowerCase();
  if (!raw) return "none";
  if (raw.includes("timeout") || raw.includes("aborted")) return "timeout";
  if (raw.includes("openai_api_key") || raw.includes("api key")) return "auth";
  if (raw.includes("http 4")) return "http-4xx";
  if (raw.includes("http 5")) return "http-5xx";
  if (raw.includes("schema")) return "schema";
  if (raw.includes("proxy") || raw.includes("fetch") || raw.includes("network")) return "proxy";
  return "other";
}

export function markAiDecisionRequest(state, channel, nowSec = Number(state?.metrics?.timeSec ?? 0)) {
  const stats = ensureAiRuntimeStats(state);
  stats.requestCount += 1;
  stats.lastRequestSec = nowSec;
  if (channel === "environment") stats.environmentRequests += 1;
  if (channel === "policy") stats.policyRequests += 1;
  return stats;
}

export function recordAiDecisionResult(state, channel, result, nowSec = Number(state?.metrics?.timeSec ?? 0)) {
  const stats = ensureAiRuntimeStats(state);
  const latencyMs = clampNonNegative(result?.latencyMs ?? 0);
  const errorMessage = String(result?.error ?? "").trim();
  const errorKind = classifyAiErrorMessage(errorMessage);
  const fallback = Boolean(result?.fallback);
  const coverageTarget = getAiCoverageTarget(state);

  stats.responseCount += 1;
  stats.lastResponseSec = nowSec;
  stats.lastLatencyMs = latencyMs;
  stats.avgLatencyMs = stats.responseCount <= 1
    ? latencyMs
    : Number((((stats.avgLatencyMs * (stats.responseCount - 1)) + latencyMs) / stats.responseCount).toFixed(2));
  stats.coverageTarget = coverageTarget;
  stats.lastResultSource = fallback ? "fallback" : "llm";
  if (channel === "environment") stats.environmentResponses += 1;
  if (channel === "policy") stats.policyResponses += 1;

  if (errorKind !== "none") {
    stats.errorCount += 1;
    stats.lastErrorKind = errorKind;
    stats.lastErrorMessage = errorMessage;
    if (errorKind === "timeout") stats.timeoutCount += 1;
  } else {
    stats.lastErrorKind = "none";
    stats.lastErrorMessage = "";
  }

  if (fallback) {
    stats.fallbackResponseCount += 1;
    if (coverageTarget === "llm") {
      stats.consecutiveFallbackResponses += 1;
      stats.lastFallbackSec = nowSec;
      const unrecoveredSec = stats.lastLiveSec > -998 ? nowSec - stats.lastLiveSec : 0;
      stats.maxUnrecoveredFallbackSec = Math.max(stats.maxUnrecoveredFallbackSec, clampNonNegative(unrecoveredSec));
    }
  } else {
    stats.llmResponseCount += 1;
    if (coverageTarget === "llm" && stats.consecutiveFallbackResponses > 0) {
      stats.recoveryCount += 1;
    }
    stats.consecutiveFallbackResponses = 0;
    stats.lastLiveSec = nowSec;
    stats.liveCoverageSatisfied = true;
  }

  if (coverageTarget !== "llm") {
    stats.consecutiveFallbackResponses = 0;
    stats.maxUnrecoveredFallbackSec = 0;
    stats.liveCoverageSatisfied = false;
  }

  return stats;
}

File: project_utopia/app/ai_runtime_stats.py
"""AI runtime telemetry (port of ``src/app/aiRuntimeStats.js``).

The JS module owns ``state.metrics.aiRuntime`` — a tally of LLM channel
requests/responses, latency, fallback behaviour, and token usage. The
Python port preserves **every JS field name** at the wire format
(camelCase string keys), but exposes a snake_case ergonomic surface on
the pydantic model.

Storage convention
------------------
Per the Phase-1 conventions doc, the runtime stats live under
``state.metrics["ai_runtime"]`` (snake_case dict key) but the *inner*
fields stay camelCase so existing NDJSON consumers keep working without
schema migration. :func:`ensure_ai_runtime_stats` enforces both invariants
(presence and clamping) and is idempotent — repeated calls do not mutate
already-valid stats.
"""

from __future__ import annotations

import math
from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

__all__ = [
    "AiRuntimeStatsModel",
    "AI_RUNTIME_FIELDS",
    "create_default_ai_runtime_stats",
    "ensure_ai_runtime_stats",
    "get_ai_coverage_target",
    "classify_ai_error_message",
    "reset_ai_runtime_stats",
]


# Ordered list of every JS field name. Kept as an immutable tuple so we can
# iterate when serializing / asserting completeness.
AI_RUNTIME_FIELDS: tuple[str, ...] = (
    "requestCount",
    "responseCount",
    "environmentRequests",
    "policyRequests",
    "environmentResponses",
    "policyResponses",
    "llmResponseCount",
    "fallbackResponseCount",
    "timeoutCount",
    "errorCount",
    "recoveryCount",
    "avgLatencyMs",
    "lastLatencyMs",
    "lastRequestSec",
    "lastResponseSec",
    "lastLiveSec",
    "lastFallbackSec",
    "maxUnrecoveredFallbackSec",
    "consecutiveFallbackResponses",
    "lastErrorKind",
    "lastErrorMessage",
    "lastResultSource",
    "coverageTarget",
    "liveCoverageSatisfied",
    # Token telemetry (S5).
    "promptTokens",
    "completionTokens",
    "cachedTokens",
    "firstTokenLatencyMs",
    "tokensPerSec",
    "kvCacheHits",
    "prefixHits",
)

# JS sentinel for "never observed". Kept as a module constant so callers
# can branch on it without magic numbers.
_NEVER_SEC: float = -999.0


def _clamp_non_negative(value: Any, fallback: float = 0.0) -> float:
    """Coerce ``value`` to a non-negative float, falling back on NaN/inf."""
    try:
        n = float(value)
    except (TypeError, ValueError):
        return fallback
    if not math.isfinite(n):
        return fallback
    return max(0.0, n)


def _finite_or(value: Any, fallback: float) -> float:
    """Return ``float(value)`` if finite, else ``fallback``."""
    try:
        n = float(value)
    except (TypeError, ValueError):
        return fallback
    if not math.isfinite(n):
        return fallback
    return n


class AiRuntimeStatsModel(BaseModel):
    """Pydantic v2 model mirroring ``state.metrics.aiRuntime``.

    The model uses ``populate_by_name=True`` with camelCase aliases so it
    can round-trip both JS-style payloads (``{"requestCount": 1}``) and
    snake_case Python access (``stats.request_count``).
    """

    model_config = ConfigDict(
        populate_by_name=True,
        extra="allow",  # permit forward-compat fields without breaking parse
    )

    request_count: int = Field(default=0, alias="requestCount")
    response_count: int = Field(default=0, alias="responseCount")
    environment_requests: int = Field(default=0, alias="environmentRequests")
    policy_requests: int = Field(default=0, alias="policyRequests")
    environment_responses: int = Field(default=0, alias="environmentResponses")
    policy_responses: int = Field(default=0, alias="policyResponses")
    llm_response_count: int = Field(default=0, alias="llmResponseCount")
    fallback_response_count: int = Field(default=0, alias="fallbackResponseCount")
    timeout_count: int = Field(default=0, alias="timeoutCount")
    error_count: int = Field(default=0, alias="errorCount")
    recovery_count: int = Field(default=0, alias="recoveryCount")
    avg_latency_ms: float = Field(default=0.0, alias="avgLatencyMs")
    last_latency_ms: float = Field(default=0.0, alias="lastLatencyMs")
    last_request_sec: float = Field(default=_NEVER_SEC, alias="lastRequestSec")
    last_response_sec: float = Field(default=_NEVER_SEC, alias="lastResponseSec")
    last_live_sec: float = Field(default=_NEVER_SEC, alias="lastLiveSec")
    last_fallback_sec: float = Field(default=_NEVER_SEC, alias="lastFallbackSec")
    max_unrecovered_fallback_sec: float = Field(default=0.0, alias="maxUnrecoveredFallbackSec")
    consecutive_fallback_responses: int = Field(default=0, alias="consecutiveFallbackResponses")
    last_error_kind: str = Field(default="none", alias="lastErrorKind")
    last_error_message: str = Field(default="", alias="lastErrorMessage")
    last_result_source: str = Field(default="none", alias="lastResultSource")
    coverage_target: Literal["fallback", "llm"] = Field(default="fallback", alias="coverageTarget")
    live_coverage_satisfied: bool = Field(default=False, alias="liveCoverageSatisfied")
    # Token telemetry (S5).
    prompt_tokens: int = Field(default=0, alias="promptTokens")
    completion_tokens: int = Field(default=0, alias="completionTokens")
    cached_tokens: int = Field(default=0, alias="cachedTokens")
    first_token_latency_ms: float = Field(default=0.0, alias="firstTokenLatencyMs")
    tokens_per_sec: float = Field(default=0.0, alias="tokensPerSec")
    kv_cache_hits: int = Field(default=0, alias="kvCacheHits")
    prefix_hits: int = Field(default=0, alias="prefixHits")

    def to_wire(self) -> dict[str, Any]:
        """Serialize to the JS-compatible camelCase dict form."""
        return self.model_dump(by_alias=True)


def create_default_ai_runtime_stats() -> dict[str, Any]:
    """Return a fresh stats dict with all 31 fields at safe defaults.

    The returned dict uses **camelCase** keys, matching the JS wire format
    consumed by NDJSON readers and the in-browser HUD (now retired).
    """
    return AiRuntimeStatsModel().to_wire()


def get_ai_coverage_target(state: dict[str, Any] | None) -> Literal["fallback", "llm"]:
    """Read ``state.ai.coverageTarget`` and normalize to one of the two literals.

    Defaults to ``"fallback"`` when ``state`` is missing the field or the
    value isn't recognized — mirrors the JS lenient parse.
    """
    if not state:
        return "fallback"
    ai = state.get("ai") if isinstance(state, dict) else None
    if not isinstance(ai, dict):
        return "fallback"
    raw = str(ai.get("coverageTarget", "")).strip().lower()
    return "llm" if raw == "llm" else "fallback"


def ensure_ai_runtime_stats(state: dict[str, Any]) -> dict[str, Any]:
    """Ensure ``state.metrics["ai_runtime"]`` exists and is well-formed.

    The function is **idempotent**: calling it twice on the same state
    leaves the contents identical (modulo ``coverageTarget`` which is
    always re-read from ``state.ai``).

    Returns
    -------
    dict[str, Any]
        The same stats dict that lives at ``state["metrics"]["ai_runtime"]``,
        so callers can chain mutations like
        ``ensure_ai_runtime_stats(state)["requestCount"] += 1``.
    """
    metrics = state.setdefault("metrics", {})
    if not isinstance(metrics, dict):  # pragma: no cover - defensive
        metrics = {}
        state["metrics"] = metrics
    stats = metrics.get("ai_runtime")
    if not isinstance(stats, dict):
        stats = create_default_ai_runtime_stats()
        metrics["ai_runtime"] = stats

    # Non-negative integer/float counters.
    for key in (
        "requestCount",
        "responseCount",
        "environmentRequests",
        "policyRequests",
        "environmentResponses",
        "policyResponses",
        "llmResponseCount",
        "fallbackResponseCount",
        "timeoutCount",
        "errorCount",
        "recoveryCount",
        "avgLatencyMs",
        "lastLatencyMs",
        "maxUnrecoveredFallbackSec",
        "consecutiveFallbackResponses",
        "promptTokens",
        "completionTokens",
        "cachedTokens",
        "firstTokenLatencyMs",
        "tokensPerSec",
        "kvCacheHits",
        "prefixHits",
    ):
        stats[key] = _clamp_non_negative(stats.get(key, 0))

    # Free-floating timestamps (may be negative sentinel _NEVER_SEC).
    for key in ("lastRequestSec", "lastResponseSec", "lastLiveSec", "lastFallbackSec"):
        stats[key] = _finite_or(stats.get(key, _NEVER_SEC), _NEVER_SEC)

    # String fields.
    stats["lastErrorKind"] = str(stats.get("lastErrorKind") or "none")
    stats["lastErrorMessage"] = str(stats.get("lastErrorMessage") or "")
    stats["lastResultSource"] = str(stats.get("lastResultSource") or "none")

    # Coverage target reflects state.ai every tick.
    stats["coverageTarget"] = get_ai_coverage_target(state)
    stats["liveCoverageSatisfied"] = bool(stats.get("liveCoverageSatisfied", False))

    # Guarantee every canonical field is present (defensive against partial
    # loads from disk).
    defaults = create_default_ai_runtime_stats()
    for key in AI_RUNTIME_FIELDS:
        if key not in stats:
            stats[key] = defaults[key]

    return stats


def reset_ai_runtime_stats(state: dict[str, Any]) -> dict[str, Any]:
    """Replace ``state.metrics["ai_runtime"]`` with a fresh default block."""
    metrics = state.setdefault("metrics", {})
    fresh = create_default_ai_runtime_stats()
    fresh["coverageTarget"] = get_ai_coverage_target(state)
    metrics["ai_runtime"] = fresh
    return fresh


def classify_ai_error_message(error_text: str = "") -> str:
    """Bucket an error string into one of the JS ``lastErrorKind`` enums."""
    raw = str(error_text or "").strip().lower()
    if not raw:
        return "none"
    if "timeout" in raw or "aborted" in raw:
        return "timeout"
    if "openai_api_key" in raw or "api key" in raw:
        return "auth"
    if "http 4" in raw:
        return "http-4xx"
    if "http 5" in raw:
        return "http-5xx"
    if "schema" in raw:
        return "schema"
    if "proxy" in raw or "fetch" in raw or "network" in raw:
        return "proxy"
    return "other"
